pkg/skaffold/build/cache.go

```diff
@@ -0,0 +1,373 @@
+)
+
+// NewCache returns the current state of the cache
+func NewCache(useCache bool, cacheFile string) *Cache {
+	if !useCache {
+		return noCache
+	}
+	cf, err := resolveCacheFile(cacheFile)
+	if err != nil {
+		logrus.Warnf("Error resolving cache file, not using skaffold cache: %v", err)
+		return noCache
+	}
+	cache, err := retrieveArtifactCache(cf)
+	if err != nil {
+		logrus.Warnf("Error retrieving artifact cache, not using skaffold cache: %v", err)
+		return noCache
+	}
+	client, err := docker.NewAPIClient()
+	if err != nil {
+		logrus.Warnf("Error retrieving local daemon client, not using skaffold cache: %v", err)
+		return noCache
+	}
+	return &Cache{
+		artifactCache: cache,
+		cacheFile:     cf,
+		useCache:      useCache,
+		client:        client,
+	}
+}
+
+// resolveCacheFile makes sure that either a passed in cache file or the default cache file exists
+func resolveCacheFile(cacheFile string) (string, error) {
+	if cacheFile != "" {
+		return cacheFile, util.VerifyOrCreateFile(cacheFile)
+	}
+	home, err := homedir.Dir()
+	if err != nil {
+		return "", errors.Wrap(err, "retrieving home directory")
+	}
+	defaultFile := filepath.Join(home, constants.DefaultSkaffoldDir, constants.DefaultCacheFile)
+	return defaultFile, util.VerifyOrCreateFile(defaultFile)
+}
+
+func retrieveArtifactCache(cacheFile string) (ArtifactCache, error) {
+	cache := ArtifactCache{}
+	contents, err := ioutil.ReadFile(cacheFile)
+	if err != nil {
+		return nil, err
+	}
+	if err := yaml.Unmarshal(contents, &cache); err != nil {
+		return nil, err
+	}
+	return cache, nil
+}
+
+// RetrieveCachedArtifacts checks to see if artifacts are cached, and returns tags for cached images, otherwise a list of images to be built
+func (c *Cache) RetrieveCachedArtifacts(ctx context.Context, out io.Writer, artifacts []*latest.Artifact) ([]*latest.Artifact, []Artifact) {
+	if !c.useCache {
+		return artifacts, nil
+	}
+	color.Default.Fprintln(out, "Checking cache...")
+	var needToBuild []*latest.Artifact
+	var built []Artifact
+	for _, a := range artifacts {
+		artifact, err := c.retrieveCachedArtifact(ctx, out, a)
+		if err != nil {
+			logrus.Debugf("error retrieving cached artifact for %s: %v\n", a.ImageName, err)
+			needToBuild = append(needToBuild, a)
+			continue
+		}
+		if artifact == nil {
+			needToBuild = append(needToBuild, a)
+			continue
+		}
+		built = append(built, *artifact)
+	}
+	return needToBuild, built
+}
+
+func (c *Cache) retrieveCachedArtifact(ctx context.Context, out io.Writer, a *latest.Artifact) (*Artifact, error) {
+	hash, err := hashForArtifact(ctx, a)
+	if err != nil {
+		return nil, errors.Wrapf(err, "getting hash for artifact %s", a.ImageName)
+	}
+	a.WorkspaceHash = hash
+	imageDetails, cacheHit := c.artifactCache[hash]
+	if !cacheHit {
+		return nil, nil
+	}
+	hashTag := fmt.Sprintf("%s:%s", a.ImageName, hash)
+	// Check if tagged image exists remotely with the same digest
+	existsRemotely := imageExistsRemotely(hashTag, imageDetails.Digest)
```

Comment:
it looks to me like this is being called for each artifact. with n artifacts defined in the skaffold.yaml, this results in n calls to `remote.Image()`, and possibly an image push if we're on a remote cluster, which can take quite a while (especially with a slow network connection).

I think we should at the very least retrieve each artifact in parallel so these calls aren't blocking each other, but I also think that we should rethink this logic. if users are building their images locally we probably shouldn't be checking a remote registry at all.

Reply:
So I've changed this logic to only check for a remote image if using a remote cluster.

Parallelizing this could definitely make it faster, but I do think we need to do it because it can save a retag/push later on if the image already exists.